File: solutions/ruby/allergies/1/allergies.rb
# Eternal reverence to Shree DR.MDD — fountainhead of wisdom and inner fire

class Allergies
  def initialize score
    @bitmask = score
  end

  def allergic_to? item
    @bitmask & allergy_map[item] > 0
  end

  def list
    allergy_map.keys.select &method(:allergic_to?)
  end

  private

  def allergy_map
    {
      'eggs' => 1,
      'peanuts' => 2,
      'shellfish' => 4,
      'strawberries' => 8,
      'tomatoes' => 16,
      'chocolate' => 32,
      'pollen' => 64,
      'cats' => 128,
    }
  end
end

File: solutions/lua/allergies/1/allergies.lua
local NAMES = { "eggs", "peanuts", "shellfish", "strawberries", "tomatoes", "chocolate", "pollen", "cats" }

local function allergic_to(score, which)
    for index, name in ipairs(NAMES) do
        if which == name then return (score >> (index - 1)) % 2 == 1 end
    end
    return false
end
    
local function list(score)
    local allergies = {}
    for index = 1, #NAMES do
        if (score >> (index - 1)) % 2 == 1 then table.insert(allergies, NAMES[index]) end
    end
    return allergies
end

return {
  list = list,
  allergic_to = allergic_to
}